packages/node/src/indexer/dictionary.service.ts

```diff
@@ -0,0 +1,156 @@
+import { SubqlEventFilter, SubqlCallFilter, MetaData } from '@subql/common';
+import fetch from 'node-fetch';
+import { ProjectIndexFilters } from './types';
+
+export type Dictionary = {
+  _metadata: MetaData;
+  batchBlocks: number[];
+  specVersions: number[];
+};
+
+@Injectable()
+export class DictionaryService implements OnApplicationShutdown {
+  private isShutdown = false;
+
+  onApplicationShutdown(): void {
+    this.isShutdown = true;
+  }
+
+  //generate dictionary query
+  private dictionaryQuery(
+    startBlock: number,
+    batchSize: number,
+    existEventHandler: boolean,
+    existExtrinsicHandler: boolean,
+    indexEvents?: SubqlEventFilter[],
+    indexExtrinsics?: SubqlCallFilter[],
+  ): string {
+    let eventFilter = ``;
+    let extrinsicFilter = ``;
+    let baseQuery = ``;
+    const metaQuery = `
+  _metadata {
+    lastProcessedHeight
+    lastProcessedTimestamp
+    targetHeight
+    chain
+    specName
+    genesisHash
+    indexerHealthy
+    indexerNodeVersion
+    queryNodeVersion
+  }`;
+    const specVersionQuery = `
+     specVersions{
+        nodes{
+          id
+          blockHeight
+        }        
+      }`;
+    baseQuery = baseQuery.concat(metaQuery, specVersionQuery);
+    if (existEventHandler && indexEvents.length !== 0) {
+      indexEvents.map((event) => {
+        eventFilter = eventFilter.concat(`
+        {
+          and:[
+          {module:{equalTo: "${event.module}"}},
+          {event:{equalTo:"${event.method}"}}
+        ]},`);
+      });
+      const eventQuery = `events(filter:{
+    blockHeight:{greaterThan:"${startBlock}"},
+    or:[
+     ${eventFilter}
+    ]
+  }, orderBy:BLOCK_HEIGHT_ASC,first: ${batchSize}){
+    nodes{
+      blockHeight
+    }
+  }`;
+      baseQuery = baseQuery.concat(eventQuery);
+    }
+    if (existExtrinsicHandler && indexExtrinsics.length !== 0) {
+      indexExtrinsics.map((extrinsic) => {
+        extrinsicFilter = extrinsicFilter.concat(`
+        {
+          and:[
+          {module:{equalTo: "${extrinsic.module}"}},
+          {call:{equalTo:"${extrinsic.method}"}}
+        ]},`);
+      });
+      const extrinsicQueryQuery = `extrinsics(filter:{
+    blockHeight:{greaterThan:"${startBlock}"},
+    or:[
+     ${extrinsicFilter}
+    ]
+  }, orderBy:BLOCK_HEIGHT_ASC,first: ${batchSize}){
+    nodes{
+      blockHeight
+    }
+  }`;
+      baseQuery = baseQuery.concat(extrinsicQueryQuery);
+    }
+    return `query{${baseQuery}}`;
+  }
+
+  async getDictionary(
+    startBlock: number,
+    batchSize: number,
+    api: string,
+    indexFilters: ProjectIndexFilters,
+  ): Promise<Dictionary> {
+    let resp;
+    const query = this.dictionaryQuery(
+      startBlock,
+      batchSize,
+      indexFilters.existEventHandler,
+      indexFilters.existExtrinsicHandler,
+      indexFilters.eventFilters,
+      indexFilters.extrinsicFilters,
+    );
+    const client = new ApolloClient({
+      cache: new InMemoryCache(),
```

Comment:
do we need to enable the cache?

Reply:
cache is not an optional parameter, but we can disable cache though 
`
defaultOptions: {
        watchQuery:{
          fetchPolicy: 'no-cache',
        },
        query:{
          fetchPolicy: 'no-cache'
        }`